### PR TITLE
feature/EUI-3615-allocate-role-page-flow

### PR DESCRIPTION
**Pull request by rayliangatos**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/EUI-3615

### Change description ###
https://tools.hmcts.net/jira/browse/EUI-3810
https://tools.hmcts.net/jira/browse/EUI-3811

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/role-access/containers/add-exclusion/add-exclusion-home/add-exclusion-home.component.ts [hunk 1]
```diff
@@ -140,7 +140,7 @@ export class AddExclusionHomeComponent implements OnInit, OnDestroy {
         break;
       }
       case ExclusionNavigationEvent.CANCEL:
-        this.router.navigateByUrl(`cases/case-details/${this.caseId}`).then(r => {
+        this.router.navigateByUrl(`cases/case-details/${this.caseId}/roles-and-access`).then(r => {
           return;
```
[the remark this answers is not in the document]
Comment (rayliangatos) [reply src/role-access/containers/add-exclusion/add-exclusion-home/add-exclusion-home.component.ts:144]: this is to suppress the ts warning here, remove it will get a ts warning.

[the remark this answers is not in the document]
Comment (udaydenduluri33) [reply src/role-access/containers/add-exclusion/add-exclusion-home/add-exclusion-home.component.ts:144]: I understand but you could add ts comment to not show the warning. Empty then is more bad

[the remark this answers is not in the document]
Comment (rayliangatos) [reply src/role-access/containers/add-exclusion/add-exclusion-home/add-exclusion-home.component.ts]: this is now removed

**Review by udaydenduluri33 — APPROVED** [review]
@rayliangatos  Thanks for allocation base functionality and also making the changes

File: api/exclusions/exclusionService.ts [hunk 2]
```diff
@@ -31,8 +31,10 @@ export async function confirmUserExclusion(req: EnhancedRequest, res: Response,
       userType: 'Judicial',
     },
   ];
-  if (['400, 401, 402, 403, 500, 503'].indexOf(req.body.exclusionDescription)) {
-    return res.status(req.body.exclusionDescription).send('error: {status: req.body.exclusionDescription}"');
+  const errorCodes: string[] = ['400', '401', '402', '403', '500', '503'];
+  const value: string = req.body.exclusionDescription;
+  if (errorCodes.indexOf(value) !== -1) {
+    return res.status(parseInt(value, 10)).send('error: {status: req.body.exclusionDescription}"');
```
Comment (connorpgpmcelroy) [inline api/exclusions/exclusionService.ts:37]: Caught an issue with this on tests - req.body.exclusionDescription is being shown as string - value should be being passed into string

File: src/role-access/components/choose-radio-option/choose-radio-option.component.ts [hunk 3]
```diff
@@ -11,7 +10,7 @@ import { OptionsModel } from '../../models/options-model';
 export class ChooseRadioOptionComponent {
 
   @Input() public title: string;
-  @Input() public caption: RoleAllocationCaptionText;
+  @Input() public caption: string;
```
Comment (connorpgpmcelroy) [inline src/role-access/components/choose-radio-option/choose-radio-option.component.ts:13]: Minor query - Why is this change necessary?

File: src/role-access/containers/allocate-role/allocate-role-home/allocate-role-home.component.ts [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+  }
+
+  public isComponentVisible(currentNavigationState: AllocateRoleState, requiredNavigationState: AllocateRoleState[]): boolean {
+    return requiredNavigationState.includes(currentNavigationState);
+  }
+
+  public onNavEvent(event: AllocateRoleNavigationEvent): void {
+    this.navEvent = {
+      event,
+      timestamp: Date.now()
```
Comment (connorpgpmcelroy) [inline src/role-access/containers/allocate-role/allocate-role-home/allocate-role-home.component.ts:73]: Minor query - Why is the timestamp necessary? Is it going to be used?

File: src/role-access/containers/allocate-role/allocate-role-navigation/allocate-role-navigation.component.ts [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+  public ngOnInit(): void {
+    this.navigationCurrentState$ = this.store.pipe(select(fromFeature.getAllocateRoleActiveState));
+  }
+
+  public isVisible(currentNavigationState: AllocateRoleState, visibleNavigationStates: AllocateRoleState[]): boolean {
+    return visibleNavigationStates.includes(currentNavigationState);
+  }
+
+  public onEventTrigger(event: AllocateRoleNavigationEvent) {
```
Comment (connorpgpmcelroy) [inline src/role-access/containers/allocate-role/allocate-role-navigation/allocate-role-navigation.component.ts:43]: Return type void

File: src/role-access/containers/allocate-role/allocate-role-search-person/allocate-role-search-person.component.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+export class AllocateRoleSearchPersonComponent implements OnInit {
+
+  @Input() public navEvent: AllocateRoleNavigation;
+
+  constructor(private readonly store: Store<fromFeature.State>) { }
+
+  public ngOnInit(): void {
+  }
+
+  public navigationHandler(navEvent: AllocateRoleNavigationEvent) {
```
Comment (connorpgpmcelroy) [inline src/role-access/containers/allocate-role/allocate-role-search-person/allocate-role-search-person.component.ts:19]: return type void

File: src/role-access/containers/allocate-role/choose-allocate-to/choose-allocate-to.component.ts [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+        optionValue: AllocateTo.RESERVE_TO_ME
+      },
+      {
+        optionId: EnumUtil(AllocateTo).getKeyOrDefault(AllocateTo.ALLOCATE_TO_ANOTHER_PERSON),
+        optionValue: AllocateTo.ALLOCATE_TO_ANOTHER_PERSON
+      }
+    ];
+  }
+
+  public navigationHandler(navEvent: AllocateRoleNavigationEvent) {
```
Comment (connorpgpmcelroy) [inline src/role-access/containers/allocate-role/choose-allocate-to/choose-allocate-to.component.ts:64]: return type void - same below

File: src/role-access/containers/allocate-role/choose-role/choose-role.component.ts [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+      {
+        optionId: EnumUtil(TypeOfRole).getKeyOrDefault(TypeOfRole.HEARING_JUDGE),
+        optionValue: TypeOfRole.HEARING_JUDGE
+      }
+    ];
+    const legalOpsOptions = [{optionId: EnumUtil(TypeOfRole).getKeyOrDefault(TypeOfRole.CASE_MANAGER), optionValue: TypeOfRole.CASE_MANAGER}];
+    this.optionsList = this.userType === PersonRole.JUDICIAL.toLowerCase() ? judicialOptions : legalOpsOptions;
+  }
+
+  public navigationHandler(navEvent: AllocateRoleNavigationEvent) {
```
Comment (connorpgpmcelroy) [inline src/role-access/containers/allocate-role/choose-role/choose-role.component.ts:70]: return type void - below as well

File: src/role-access/models/allocate-role-navigation-event.enum.ts [hunk 9]
```diff
@@ -0,0 +1,6 @@
+export enum AllocateRoleNavigationEvent {
+  BACK,
+  CANCEL,
+  CONTINUE,
+  CONFIRM,
+}
```
Comment (connorpgpmcelroy) [inline src/role-access/models/allocate-role-navigation-event.enum.ts:6]: Minor query - Do we need delete here?

**Review by connorpgpmcelroy — CHANGES_REQUESTED** [review]
Small changes necessary but mostly looks good

File: src/role-access/components/choose-radio-option/choose-radio-option.component.ts (hunk 3 above)
Comment (rayliangatos) [reply src/role-access/components/choose-radio-option/choose-radio-option.component.ts:13]: this is to make caption to be dynamic, it's not only for specific enum.

File: src/role-access/models/allocate-role-navigation-event.enum.ts (hunk 9 above)
Comment (rayliangatos) [reply src/role-access/models/allocate-role-navigation-event.enum.ts:6]: Nope I think, as the Delete is in different user journey, but you could put Delete here however the navigation doesn't have Delete state in the single page.

File: src/role-access/containers/allocate-role/allocate-role-home/allocate-role-home.component.ts (hunk 4 above)
Comment (rayliangatos) [reply src/role-access/containers/allocate-role/allocate-role-home/allocate-role-home.component.ts:73]: it's not used in the business requirement, but technically this will tell when the user trigger the event.

File: api/exclusions/exclusionService.ts (hunk 2 above)
Comment (rayliangatos) [reply api/exclusions/exclusionService.ts]: good spot, this was originally from Paul H, I am not aware this template is doing nothing. Changed now.

File: src/role-access/containers/allocate-role/allocate-role-navigation/allocate-role-navigation.component.ts (hunk 5 above)
Comment (rayliangatos) [reply src/role-access/containers/allocate-role/allocate-role-navigation/allocate-role-navigation.component.ts]: added

File: src/role-access/containers/allocate-role/allocate-role-search-person/allocate-role-search-person.component.ts (hunk 6 above)
Comment (rayliangatos) [reply src/role-access/containers/allocate-role/allocate-role-search-person/allocate-role-search-person.component.ts]: added

File: src/role-access/containers/allocate-role/choose-allocate-to/choose-allocate-to.component.ts (hunk 7 above)
Comment (rayliangatos) [reply src/role-access/containers/allocate-role/choose-allocate-to/choose-allocate-to.component.ts]: added

File: src/role-access/containers/allocate-role/choose-duration/choose-duration.component.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+export class ChooseDurationComponent implements OnInit {
+
+  @Input() public navEvent: AllocateRoleNavigation;
+
+  constructor(private readonly store: Store<fromFeature.State>) { }
+
+  public ngOnInit(): void {
+  }
+
+  public navigationHandler(navEvent: AllocateRoleNavigationEvent) {
```
[the remark this answers is not in the document]
Comment (rayliangatos) [reply src/role-access/containers/allocate-role/choose-duration/choose-duration.component.ts]: added

File: src/role-access/containers/allocate-role/choose-role/choose-role.component.ts (hunk 8 above)
Comment (rayliangatos) [reply src/role-access/containers/allocate-role/choose-role/choose-role.component.ts]: added

File: src/role-access/components/choose-radio-option/choose-radio-option.component.ts (hunk 3 above)
Comment (connorpgpmcelroy) [reply src/role-access/components/choose-radio-option/choose-radio-option.component.ts:13]: ok sure

File: src/role-access/models/allocate-role-navigation-event.enum.ts (hunk 9 above)
Comment (connorpgpmcelroy) [reply src/role-access/models/allocate-role-navigation-event.enum.ts:6]: ok sure

File: src/role-access/containers/allocate-role/allocate-role-home/allocate-role-home.component.ts (hunk 4 above)
Comment (connorpgpmcelroy) [reply src/role-access/containers/allocate-role/allocate-role-home/allocate-role-home.component.ts:73]: Ok sure - I suppose we could use it in the future

**Review by connorpgpmcelroy — APPROVED** [review]
Comments addressed - Looks good to me
